rfcs/0009_Unified_Static_Memory_Planning.md

```diff
@@ -0,0 +1,473 @@
+    }
+```
+```
+// metadata.h
+
+    #define TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE xxxx
+
+    typedef struct {
+       uint8_t* dtcm;
+       uint8_t* sram;
+    }  TVMWorkspaces_my_model;
+
+    typedef struct {
+       uint8_t* itcm;
+       uint8_t* flash;
+    }  TVMParameters_my_model;
+
+    typedef struct {
+       uint8_t* input0;
+    }  TVMInputs_my_model;
+
+    typedef struct {
+       uint8_t* output0;
+    }  TVMOutputs_my_model;
+```
+### User Application
+```
+    // The User Application 
+        extern  const TVMModel my_model;
+        __attribute__((section( "ITCM" )  const uint8_t   my_model_params_1[TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE] = <param_1_data>;
+        __attribute__((section( "FLASH" ), aligned( 16 )))  const uint8_t my_model_params_2[TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE] = <param_2_data>;
+        __attribute__((section( "DTCM" )  static uint8_t workspace_buffer_1[TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE];
+        __attribute__((section( "SRAM" ), aligned( 16 )))  static uint8_t workspace_buffer_2[TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE];
+
+    int main(...) {
+         ...
+         TVMContext context;
+         TVMInputs_my_model_1 inputs = {input};
+         TVMOutputs_my_model_1 outputs = {output};
+         TVMWorkspaces_my_model workspaces = {
+             .sram = &workspace_buffer_1,
+             .dtcm = &workspace_buffer_2,
+         };
+         TVMParameters_my_model parameters = {
+             .flash = &my_model_params_1,
+             .itcm = &my_model_params_2
+         };
+         TVMSetWorkspaces(&context, &workspaces);
+         TVMSetParameters(&context, parameters);
+         TVMExecute(&my_model, &inputs, &outputs, &context);
+    }
+```
+# Reference-level explanation
+
+## Overview
+
+This should be a IRModule (TIR) → IRModule (TIR) pass.
+
+Inputs : 
+* AoT TIR PrimFunc ( the control function describing the call graph to operators)
+* All Operator Functions
+* the maximum size for each pool We could use "pinned_memory" (see below) to tag buffers with suggested priority order determined by the scheduler.
+
+The idea is USMP will try to pool them using the preferred "pinned_memory" and fallback whenever the size is exceeding the user provided max size for each pool (if any)
+
+Outputs : 
+* AoT TIR PrimFunc accepting pool buffers from the user.
+* All Operator functions accepting pool buffers.
+  * Each operator function should address using the correct offset in the correct pool buffer
+
+Special Parametric Inputs : 
+* function : The algorithm to be used for planning From a component PoV, the algorithm is a special input with a defined interface.
+
+The current proposal for the interface of the memory planning algorithm is as follows :
+```
+    struct BufferInfo {
```

Comment:
It makes sense and aligns with our thinking as well.

This is something that will be passed from the user from the TVMC interface. However, TVM currently is lacking support to hold compiler(-wide) flags where we currently end up posting them as target_host attrs. During the runtime of the compiler, these values would be global constants, therefore we are kind of discussing (internally) best way to represent them generically in an extensible way when more such compiler-flags (e.g. --executor is another example here) are added (e.g., possibly as some sort of AttrRegistry) and hold them during a compilation run.

Other reason for not defining them here is that the TVM's schedulers may would like to have access them in future. 

If you have any plans/ideas to this effect, we'd love to hear them :) 

cc : @leandron @Mousius

Reply:
cc @jroesch regarding compiler input and output interfaces.

Reply:
I am trying to write down some guidelines on where to store these things, my 2c is that we should add Attributes to the IRModule to also use as a per compilation place to store global data instead of keying them off target. Is there any reason this wouldn't work for you? I think we should move in this direction so all the useful state for compilation is readable directly from the compilation unit.

Reply:
I agree with @jroesch on this, I think that it would be great to try to stope the information in the IRModule attrs. This would also be consistent with some of the cleanups we are currently doing -- I am starting to try to push the info stored in the TargetMap data structure into function attributes.

Reply:
Hi @jroesch @electriclilies,

Yes, that sounds like a good idea.

Only concern is these are constant inputs in the compilation flow -- that are not supposed to get mutated in passes.
But, yea it could still work and the passes will not need anything more than IRModule -- that seems attractive. Let me know if you have anymore thoughts around this.

If we all seem to agree, yes we can introduce the 'PoolInfo' structure here and add as an Attribute to the IRModule.